openssh_keypair - Add password protected key check

MaxBab opened the pull request:
##### SUMMARY
The ssh key may be created manually prior the task execution with a
passphrase. And the task will be executed on the same key.

The module will check the private key and if the key is password
protected, the task will fail with the following message:
"The key is protected with a passphrase. Unable to proceed."

"Fixes #63910"

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
openssh_keypair

MaxBab committed:
openssh_keypair - Add logic to handle password protected or broken key

* The ssh key may be created manually prior the task execution with a
  passphrase. And the task will be executed on the same key.
* The ssh key may be broken and not usable.

The module will check the private key and if the key is password
protected or broken, it will be overridden.
The check of the ssh key performed by retrieve the public key from the
private key.

Set the "self.force" check before the "isPrivateKeyValid" check.
In case of any issue with the "isPrivateKeyValid" function, the user
will be able to force the regeneration of the key with the "force: yes"
argument.

felixfontein commented on test/integration/targets/openssh_keypair/tasks/main.yml:
```diff
@@ -79,4 +79,13 @@
+  openssh_keypair:
+    path: '{{ output_dir }}/privatekey8'
+  register: privatekey8_result
+  ignore_errors: true
```
Maybe add another test that in case `force: yes` is specified, the module won't fail.

felixfontein commented:
@MaxBab did you have any chance to look at the failing CI tests so far? It looks like that for some versions of `ssh-keygen`, your change cannot distinguish between a totally broken file (in this case, a file of zero bytes) and a passphrase-protected file.

MaxBab commented:
@felixfontein 
Yes, I have noticed that the problem in the version of ssh.
I wrote the fix on the fedora os and the version of the openssh - openssh-7.9p1-6.
After I looked at the CI results I tested the fix on the rhel 7.7 os and it indeed failed. The version of openssh is - openssh-7.4p1-16.

The problem between the versions is that in the older version, the command I'm trying to get the error message with, still ask for the password and ignores the "SSH_ASKPASS=/bin/false" I'm using (line 244).

Right now, I can't think of any other way to check if the key is protected with the password.
I can't base my decision on the empty key because it could be a broken key and as a result should be regenerated.

I'm looking for a way to skip the password ask as it works in the new version and catch the error message that will tell that the password is incorrect.

I'm still looking for some way to do it, but currently, it keeps stuck on the password prompt.
Do you have any ideas, please?

MaxBab commented:
@felixfontein 
I have fixed the CI failed tests.
I changed the logic of the module regarding the broken keys in addition to the password protected.

The module will check the private key and if the key is password
protected or broken, the task will fail with the following message:
"Unable to read the key. The key is protected with a passphrase or broken. Unable to proceed."
The check of the ssh key performed by retrieves the public key from the private key.

In order to regenerate the broken key, "force" flag should be used.
The module changes the logic of execution with broken key.
Before the change, the broken key was regenerated without any ask.
Now, the broken key will be regenerated only if "force" flag is specified.

felixfontein commented:
@MaxBab 

> The module will check the private key and if the key is password
> protected or broken, the task will fail with the following message:
> "Unable to read the key. The key is protected with a passphrase or broken. Unable to proceed."
> The check of the ssh key performed by retrieves the public key from the private key.
> 
> In order to regenerate the broken key, "force" flag should be used.
> The module changes the logic of execution with broken key.
> Before the change, the broken key was regenerated without any ask.
> Now, the broken key will be regenerated only if "force" flag is specified.

This is different than what all other crypto modules do: they regenerate automatically for broken keys (or if password doesn't match). I'm not sure whether it is a good idea to change the behavior in this way. How about simply re-generating the key in this case? Clearly the key isn't what the user expects (or should expect :D) from the module.

Are there other opinions on this?

MaxBab commented:
@felixfontein 

Let me describe to you the flow of the password-protected key check and how it involved the same behavior on the broken key as well.
Maybe we could think of a way to work around it and keep the broken key behavior flow untouched.

The old format of the ssh keys has an "ENCRYPTED" value within the private key, specifying that the private key is encrypted with a password.
The new format of the keys doesn't have this indication.

As a result, the only way for me to check if the ssh key is password-protected is to try to retrieve the public key from the private key by the following command: "ssh-keygen -P '' -yf /tmp/sshkey".
In the case of the password-protected key, I get a message, that the passphrase is incorrect. In the other case, I get the public key printed.

That's my way of checking the key.
What happens with the broken key, is that I actually get the same error telling me that the passphrase is incorrect. I have no other way to check if the key is broken. At least I didn't find it.

felixfontein commented:
Ok, so we in general cannot distinguish between broken and passphrase protected. So how about always regenerating when one of these is the case? That solution will work in any case, and the module would behave the same as the other crypto modules.

MaxBab commented:
@felixfontein 
Agree.
Updated the patch.
Also, I added a note about this.

felixfontein reviewed (approved):
Looks good!

Could you please re-add the logic switch here? https://github.com/ansible/ansible/compare/02509801d9c7ed724435317344e6ee6d396f38ec..4fbdd671c4bf47f3075e32dd92c25a4f616544fd#diff-438137003e3ebd7c7c25344bd149a95eL174 (i.e. check `self.force` before calling `self.isPrivateKeyValid()`) This will make the module more future-proof. (In case there ever is another problem with `isPrivateKeyValid`, users can at least still force regeneration with `force: yes` instead of having to resort to calling `ssh-keygen` manually.)

MaxBab commented:
Done.

felixfontein reviewed (approved):
LGTM

felixfontein commented:
@MaxBab thanks a lot for fixing this! I'll create backports so the fix will find its way into 2.8.x and 2.9.x as well.

MaxBab commented:
Cool.
Thanks for the help.
